packages/terra-utils/lib/index.ts

```diff
@@ -0,0 +1,94 @@
+import { Block, SwapParams, Transaction, terra } from '@liquality/types'
+import { addressToString, validateExpiration, validateSecretHash, validateValue } from '@liquality/utils'
+import { InvalidAddressError } from '@liquality/errors'
+import { DateTime } from 'luxon'
+
+export const normalizeBlock = (data: any): Block => ({
+  hash: data.block_id.hash,
+  timestamp: convertDateToTimestamp(data.block.header.time),
+  size: Number(data.block.header.height),
+  number: Number(data.block.header.height),
+  parentHash: data.block.last_commit.block_id.hash
+})
+
+export const normalizeTransaction = (
+  data: any,
+  asset: string,
+  currentBlock?: number
+): Transaction<terra.InputTransaction> => {
+  const msg = data.tx.msg?.[0] || data.tx.value?.msg?.[0]?.value
+
+  let value = 0
+
+  if (Array.isArray(msg?.init_coins)) {
+    value = msg.init_coins.find((e: any) => e.denom === asset)?.amount
+  } else if (typeof msg?.init_coins === 'object') {
+    value = msg.init_coins.get(asset)?.amount
+  }
+
+  let txParams = msg?.init_msg || msg?.execute_msg || {}
+
+  if (Object.keys(txParams).length) {
+    const initMsg = msg?.init_msg
+    const executeMsg = msg?.execute_msg
+
+    if (initMsg) {
+      txParams = initMsg
+    }
+
+    if (executeMsg) {
+      txParams.method = executeMsg
+
+      if (txParams.method.claim) {
+        txParams.secret = txParams.method.claim.secret
+      }
+    }
+  }
+
+  const logs = data.logs?.[0]
+
+  const contractAddress =
+    logs?.eventsByType?.execute_contract?.contract_address[0] ||
+    logs?.events?.find((e: any) => e.type === 'wasm')?.attributes.find((e: any) => e.key === 'contract_address')
+      .value ||
+    ''
+
+  return {
+    value: Number(value),
+    hash: data.txhash,
+    confirmations: Math.min(currentBlock - data.height, 10),
+    ...(txParams?.secret && { secret: txParams.secret }),
+    _raw: {
```

Comment:
The _raw part of the transaction is supposed to be the actual raw response from the RPC
I think if you want to parse it to get other details, you should write a separate function that does that

Reply:
Done